src/org/kms/core/KMSUtils.java

```diff
@@ -0,0 +1,32 @@
+package org.kms.core;
+
+import java.util.HashSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class KMSUtils {
+
+	private static HashSet<String>lockSet = new HashSet<String>();
+	private static Lock lock = new ReentrantLock();
+
+	static HashSet<String> getMap(){
+		return lockSet;
+	}
+
+	public static boolean putWithLock(String id){
+		if (lockSet.contains(id)){
+			return false;
```

Comment:
In case if a thread comes just after the lock is acquired by a thread with request for same user id, there is no need to acquire lock for the next thread and it can directly go to while loop. That is why I have added a check there